File: turtlesim_openloop_control/src/turtlesim_openloop_control_node.cpp
// This work is sponsored by the Department of the Air Force under Air Force
// Contract FA8721-05-C-0002. Opinions, interpretations, conclusions, and
// recommendations are those of the author and are not necessarily endorsed by
// the United States Government.

#include "ros/ros.h"
#include "geometry_msgs/Twist.h"

class TurtlesimOpenLoopNode
{
public:
  TurtlesimOpenLoopNode();

private:
  ros::NodeHandle nh_;
  ros::Publisher cmd_vel_pub_;
  ros::Timer timer_;

  double period_;
  unsigned counter_; // state variable

  void timerCallback(ros::TimerEvent const& event);
};

TurtlesimOpenLoopNode::TurtlesimOpenLoopNode() :
  counter_(0)
{
  cmd_vel_pub_ = nh_.advertise<geometry_msgs::Twist>("cmd_vel", 1);

  // create a periodic timer at a fixed rate
  period_ = 0.1; // 10 Hz
  timer_ = nh_.createTimer(ros::Duration(period_), &TurtlesimOpenLoopNode::timerCallback, this);
}

void TurtlesimOpenLoopNode::timerCallback(ros::TimerEvent const& event)
{
  geometry_msgs::Twist cmd_vel_msg;
  cmd_vel_msg.linear.x = 0;
  cmd_vel_msg.linear.y = 0;
  cmd_vel_msg.linear.z = 0;
  cmd_vel_msg.angular.x = 0;
  cmd_vel_msg.angular.y = 0;
  cmd_vel_msg.angular.z = 0;

  /* go forward for 3 seconds (30 counts at 10hz) at 1.1 meters/second,
   * then turn at 45 deg/sec for 2 seconds (20 counts) */
  if (counter_ < 30)
    cmd_vel_msg.linear.x = 1.1;
  else
    cmd_vel_msg.angular.z = 0.785398163;

  // check for end of cycle
  counter_++;
  if (counter_ >= 50)
    counter_ = 0;

  cmd_vel_pub_.publish(cmd_vel_msg);
}

int main(int argc, char *argv[])
{
  ros::init(argc, argv, "turtlesim_openloop_control_node");
  TurtlesimOpenLoopNode node;
  ros::spin();
  return 0;
}

File: turtlesim_openloop_control/scripts/turtlesim_openloop_control_node.py
#!/usr/bin/python
#

# This work is sponsored by the Department of the Air Force under Air Force
# Contract FA8721-05-C-0002. Opinions, interpretations, conclusions, and
# recommendations are those of the author and are not necessarily endorsed by
# the United States Government.

import rospy
from geometry_msgs.msg import Twist

class TurtlesimOpenLoopNode:
    def __init__(self):
        # state variables
        self.counter = 0

        # create a periodic interrupt at a fixed rate
        self.period = 0.1 # 10 Hz
        rospy.Timer(rospy.Duration(self.period), self.timer_callback)

        self.cmd_vel_pub = rospy.Publisher("cmd_vel", Twist)

    def timer_callback(self, event):
        cmd_vel_msg = Twist()
        cmd_vel_msg.linear.x = 0
        cmd_vel_msg.linear.y = 0
        cmd_vel_msg.linear.z = 0
        cmd_vel_msg.angular.x = 0
        cmd_vel_msg.angular.y = 0
        cmd_vel_msg.angular.z = 0
        
        # go forward for 3 seconds (30 counts at 10hz) at 1.1
        # meters/second, then turn at 45 deg/sec for 2 seconds (20
        # counts)
        if self.counter < 30:
            cmd_vel_msg.linear.x = 1.1
        else:
            cmd_vel_msg.angular.z = 0.785398163

        # check for end of cycle
        self.counter += 1
        if self.counter >= 50:
            self.counter = 0

        # publish the velocity command message
        self.cmd_vel_pub.publish(cmd_vel_msg)

if __name__ == "__main__":
    rospy.init_node("turtlesim_openloop_control_node")
    node = TurtlesimOpenLoopNode()
    rospy.spin()
